File: csharp/utils/Visualizer.cs
using SpiderNavigation.Models;

namespace SpiderNavigation.Utils
{
    public static class Visualizer
    {
        public static void DisplayPath(List<Position> path, int gridWidth, int gridHeight)
        {
            Console.WriteLine("\n🕸️  NAVIGATION PATH VISUALIZATION:\n");
            
            // Create grid
            var grid = new char[gridHeight + 1, gridWidth + 1];
            for (int y = 0; y <= gridHeight; y++)
                for (int x = 0; x <= gridWidth; x++)
                    grid[y, x] = '·';
            
            // Mark path
            for (int i = 0; i < path.Count; i++)
            {
                var step = path[i];
                var symbol = i == path.Count - 1 ? '★' : GetDirectionSymbol(step.Orientation);
                grid[step.Y, step.X] = symbol;
            }
            
            // Display grid (inverted Y-axis for proper visualization)
            for (int y = gridHeight; y >= 0; y--)
            {
                for (int x = 0; x <= gridWidth; x++)
                {
                    Console.Write(grid[y, x] + " ");
                }
                Console.WriteLine();
            }
            
            Console.WriteLine("\n📈 PATH STEPS:");
            for (int i = 0; i < path.Count; i++)
            {
                Console.WriteLine($"Step {i}: ({path[i].X}, {path[i].Y}) facing {path[i].Orientation}");
            }
        }
        
        private static char GetDirectionSymbol(string orientation) => orientation switch
        {
            "Up" => '↑',
            "Right" => '→', 
            "Down" => '↓',
            "Left" => '←',
            _ => '·'
        };
    }
}

File: nodejs/src/utils/Visualizer.js
class Visualizer {
    static displayPath(path, gridWidth, gridHeight) {
        console.log('\n🕸️  NAVIGATION PATH VISUALIZATION:\n');
        
        // Create grid
        const grid = [];
        for (let y = 0; y <= gridHeight; y++) {
            const row = [];
            for (let x = 0; x <= gridWidth; x++) {
                row.push('·');
            }
            grid.push(row);
        }
        
        // Mark path
        path.forEach((step, index) => {
            const symbol = this.getDirectionSymbol(step.orientation);
            grid[step.y][step.x] = index === path.length - 1 ? '★' : symbol;
        });
        
        // Display grid (inverted Y-axis for proper visualization)
        for (let y = grid.length - 1; y >= 0; y--) {
            let row = '';
            for (let x = 0; x < grid[y].length; x++) {
                row += grid[y][x] + ' ';
            }
            console.log(row);
        }
        
        console.log('\n📈 PATH STEPS:');
        path.forEach((step, index) => {
            console.log(`Step ${index}: (${step.x}, ${step.y}) facing ${step.orientation}`);
        });
    }
    
    static getDirectionSymbol(orientation) {
        const symbols = { 'Up': '↑', 'Right': '→', 'Down': '↓', 'Left': '←' };
        return symbols[orientation] || '·';
    }
}

module.exports = { Visualizer };